### PR TITLE
add PowerDownBotManager

### PR DESCRIPTION
**Pull request by dnqbob**
Make AI cut down power if low power, and restore later if possible.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		[Desc("Delay (in ticks) between toggle")]
+		public readonly int Interval = 150;
+
+		[Desc("ConditionToken that powerdown")]
+		public readonly string PowerdownCondition = "powerdown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotMoudle(init.Self, this); }
+	}
+
+	public class PowerDownBotMoudle : ConditionalTrait<PowerDownBotMoudleInfo>, IBotTick
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Module

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Also, 2019 -> 2020

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: ~..........That is too embrassing~

That is even embarrassing

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: ok

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+		public void BotTick(IBot bot)
+		{
+			if (toggleTick > 0 || playerPower == null)
+			{
+				toggleTick--;
+				return;
+			}
+
+			var toggleableBuildings = GetToggleableBuildings(bot);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: This looks needlessly complicated. Is there a reason for not simply (a) returning an `IEnumerable<TraitPair<Power>>` here, filtered to just things that can be powered down and sorted by largest-least power drain, and (b) keeping a separate `List<TraitPair<Power>>` on the object to track what has been powered down to restore later?

File: mods/cnc/rules/ai.yaml [hunk 4]
```diff
@@ -307,3 +307,5 @@ Player:
 			orca: 10
 		UnitLimits:
 			harv: 8
+	PowerDownBotMoudle:
```
Comment (pchote) [inline mods/cnc/rules/ai.yaml]: TD doesn't support powering down structures.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			if (player == null || player.PlayerActor == null)
+			{
+				playerPower = null;
+				return;
+			}
+
+			playerPower = player.PlayerActor.TraitOrDefault<PowerManager>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: This should be assigned from `Created` instead - see `HarvesterBotModule` initializing `requestUnitProduction` for an example.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Because it may be fast,  I guess.  By using `SortedList`,  I can skip those who cause same power change when searching through (it is possible for AI to build a lot of same type of defences) meanwhile `SortedList` can sort those classified group in order.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+			if (world.Type == WorldType.Editor)
+				return;
+
+			player = self.Owner;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			if (player == null || player.PlayerActor == null)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: `player` can never be null, can it? `PlayerActor` should never be null?

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI powerdown.")]
+	public class PowerDownBotMoudleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between toggle")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: `Delay (in ticks) between toggling powerdown.`

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 1 above)
Comment (RobFog) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: The file is still called PowerDownBotMoudle.cs?

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: This should also be only generated when we actually need to power anything down. Keeping a list of what has been powered down will save that lookup as well and thus be faster.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: > track what has been powered down to restore later? 

As for this, the owner can change and the actor can die, if AI captures a powerdown building this track will not affect this new building. I think rescan can be better at this. I get a better idea on optimization, though. I can scan due to the need of upper calling.

File: mods/cnc/rules/ai.yaml (hunk 4 above)
Comment (dnqbob) [reply mods/cnc/rules/ai.yaml]: Thanks for mentioning.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 5 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Thanks for guide. I initialize it by using the way of `BaseBuilderBotModule` but it is buggy at shellmap

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: 1) Why would you need to skip over objects when powering down?

2) Capturing isn't enabled in most mods, and in any case we should really be restoring power to a building if it changes owner. Fixing this would make the code here much simpler.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 6 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: On shellmap without this null check will crash the game

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 7 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Thanks

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: > Why would you need to skip over objects when powering down?

For example, we have N x -40 defences (may be different) needs to be turn on but we only have 70 extra power for that. By using my idea here we turn on only one of the -40 defences, skip the rest and search -20 or -10 groups that has already in list.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Yeah but changed in my version. still things to fix.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: That adds a significant complexity overhead for I expect little observable benefit ingame. Can we please use a simpler approach to keep the code understandable and maintainable for the future?

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Another problem. On trace, how can we find out the units that are dead/ owner changed meanwhile delete them safely and less complexly? I used to make a trace in fact but it give me "Collection was modified; enumeration operation may not execute" in `foreach`. I know I have to use `for` for this but I think it will also be complex to look at. 

In this case I believe rescan is better. and thus we don't need to fix toggle on owner change. since toggle condition on order can be used not only powerdown.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Take a look at `CaptureManagerBotModule` and `HarvesterManagerBotModule` to see how to clear stale actor references.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 6 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: If that is happening here but nowhere else then its a good sign that the problem is somewhere else, and should be fixed there rather than worked around here.

Keeping standard conventions across the traits is important for maintainability so custom workarounds aren't good enough!

**Comment by MlemandPurrs** [comment]
i will be useless nitpicking: Module not Moudle (dont forget to correct typo)

**Comment by dnqbob** [comment]
should be all fixed.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 6 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: Use `created` can solve this. However, why `BaseBuilderBotModule` use `playerPower = player.PlayerActor.TraitOrDefault<PowerManager>();` in `TraitEnabled()` but works out good? Will it also buggy when remove the line of "RequiresCondition"?

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: fixed

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: fixed

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: fixed.  but captured toggled building will now not listen to the order.

File: mods/cnc/rules/ai.yaml (hunk 4 above)
Comment (dnqbob) [reply mods/cnc/rules/ai.yaml]: fixed

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs (hunk 7 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs]: fixed

**Comment by dnqbob** [comment]
emmmm, hello? Anyone here?

**Comment by dnqbob** [comment]
> Capturing isn't enabled in most mods, and in any case we should really be restoring power to a building if it changes owner. Fixing this would make the code here much simpler.

I make a fix here but need you to test it, especially on AI capture use case.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			toggledBuildings = new List<BuildingPowerWrapper>();
+		}
+
+		int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTrait = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
```
Comment (phrohdoh) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
			var powerTraits = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
```

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+			var powerMulTrait = a.TraitsImplementing<PowerMultiplier>().ToArray();
```
Comment (phrohdoh) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
			var powerMulTraits = a.TraitsImplementing<PowerMultiplier>().ToArray();
```

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+			// When there is extra power, check if AI can toggle on
+			if (power > 0)
+			{
+				foreach (var bpw in toggledBuildings)
+				{
+					if (power + bpw.PowerChanging < 0)
+						continue;
+
+					bot.QueueOrder(new Order("PowerDown", bpw.Actor, false));
```
Comment (phrohdoh) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: I think we want something other than `PowerDown` here, else the AI will only disable.

Comment (phrohdoh) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: Do we want to remove the current iteration's item from `toggledBuildings` (after the loop)?

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+		{
+			var toggleable = bot.Player.World.ActorsHavingTrait<ToggleConditionOnOrder>(t => !t.IsTraitDisabled && !t.IsTraitPaused)
+				.Where(a => a != null && !a.IsDead && a.Owner == player && a.Info.HasTraitInfo<PowerInfo>() && a.Info.HasTraitInfo<PowerMultiplierInfo>() && a.Info.HasTraitInfo<BuildingInfo>());
+
+			return toggleable;
+		}
+
+		IEnumerable<BuildingPowerWrapper> GetOnlineBuildings(IBot bot)
+		{
+			List<BuildingPowerWrapper> toggleableBuilding = new List<BuildingPowerWrapper>();
```
Comment (phrohdoh) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
			List<BuildingPowerWrapper> toggleableBuildings = new List<BuildingPowerWrapper>();
```

**Review by phrohdoh — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: Hi here.


> I think we want something other than `PowerDown` here, else the AI will only disable.

In fact `PowerDown` is a toggle function order, which means you can use `PowerDown` to turn on/off when building is off/on. It is indeed confusing at name.



> Do we want to remove the current iteration's item from `toggledBuildings` (after the loop)?

It is other OpenRA member required to add this list to keep tracking the toggled buildings of AI for performance. If it is not that time taking I also want a rescan here and left `Toggle-ableCondition` trait alone.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 11 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: typo fixed.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 9 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: typo fixed.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 8 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: typo fixed.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: However, since "powerdown" is used as a ... common sense in OpenRA so I would make it unchanged until there will be someone change this common sense here.

**Comment by dnqbob** [comment]
Rebase to newest commits to checkout conflicts

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+		int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTraits = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+			var powerMulTraits = a.TraitsImplementing<PowerMultiplier>().ToArray();
+			if (powerTraits.Any())
+			{
+				powerChangingIfToggled = powerTraits.Sum(p => p.Info.Amount) * (powerMulTraits.Sum(p => p.Info.Modifier) - 100) / 100;
+				if (powerMulTraits.Where(t => !t.IsTraitDisabled).Any())
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
				if (powerMulTraits.Any(t => !t.IsTraitDisabled))
```

**Review by Mailaender — DISMISSED** [review]
This seems to work as promised.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+			// Special case handling is required for the Player actor.
+			// Created is called before Player.PlayerActor is assigned,
+			// so we must query player traits from self, which refers
+			// for bot modules always to the Player actor.
+			playerPower = self.TraitOrDefault<PowerManager>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
			toggleTick = world.LocalRandom.Next(Info.Interval);
```

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 12 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: ok, but I need to change that after 8 hours.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+			isToggledBuildingsValid = a => a.Owner == self.Owner && !a.IsDead && a.IsInWorld && GetTogglePowerChanging(a) < 0;
+		}
+
+		protected override void Created(Actor self)
+		{
+			// Special case handling is required for the Player actor.
+			// Created is called before Player.PlayerActor is assigned,
+			// so we must query player traits from self, which refers
+			// for bot modules always to the Player actor.
+			playerPower = self.TraitOrDefault<PowerManager>();
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]:
```suggestion
			playerPower = self.Owner.PlayerActor.TraitOrDefault<PowerManager>();
```

since https://github.com/OpenRA/OpenRA/pull/18182

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 12 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: fixed

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 13 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: fixed

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 14 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: fixed

**Comment by dnqbob** [comment]
Rebase to newest bleed to checkout conflicts

File: OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs [hunk 15]
```diff
@@ -103,5 +103,12 @@ protected override void TraitResumed(Actor self)
 			// Unpausing the trait restores the previous state
 			SetCondition(self, enabled);
 		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			// Switch toggle back to untoggled if ownerchanged
+			enabled = false;
+			SetCondition(self, false);
+		}
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: Won't the condition be removed from `TraitDisabled`? Or does removing an actor from the world not trigger `TraitDisabled`?

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+		{
+			toggleTick = world.LocalRandom.Next(Info.Interval);
+			toggledBuildings = new List<BuildingPowerWrapper>();
+		}
+
+		int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTraits = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+			var powerMulTraits = a.TraitsImplementing<PowerMultiplier>().ToArray();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: This can be moved behind the `if` check to save us a lookup.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+		public override object Create(ActorInitializer init) { return new PowerDownBotModule(init.Self, this); }
+	}
+
+	public class PowerDownBotModule : ConditionalTrait<PowerDownBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		PowerManager playerPower;
+		int toggleTick;
+		readonly Func<Actor, bool> isToggledBuildingsValid;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs:37]: It would be better to make this `Func<BuildingPowerWrapper, bool>` and remove the `GetTogglePowerChanging` call.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (toggleTick > 0 || playerPower == null)
+			{
+				toggleTick--;
+				return;
+			}
+
+			var power = playerPower.ExcessPower;
+			toggledBuildings = toggledBuildings.Where(bpw => isToggledBuildingsValid(bpw.Actor)).OrderByDescending(bpw => bpw.PowerChanging).ToList();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: We should only do this in the `power > 0` case when we actually need the result.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+			else if (power < 0)
+			{
+				var buildingsCanBeOff = GetOnlineBuildings(bot);
+				foreach (var bpw in buildingsCanBeOff)
+				{
+					if (power > 0)
+						break;
+
+					bot.QueueOrder(new Order(Info.OrderName, bpw.Actor, false));
+					toggledBuildings.Add(new BuildingPowerWrapper(bpw.Actor, -bpw.PowerChanging));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: Can we even cache `PowerChanging` reliably here, considering the multipliers might change?

**Review by abcdefg30 — COMMENTED** [review]
Can't we use `INotifyPowerLevelChanged` instead of `Tick`?

File: OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs (hunk 15 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: The owner change is to solve a problem that AI-captured building which has already been in power down state won't be able to controlled by this function since we only restore the power consumption in cached list. 

> Capturing isn't enabled in most mods, and in any case we should really be restoring power to a building if it changes owner. Fixing this would make the code here much simpler.

by @pchote 

So I just did that.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 16 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: good one.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 18 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: good point. It seems some place in this module is still crude and disappointing.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 17 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs:37]: But I guess it will make code more complex here, meanwhile `GetTogglePowerChanging` is a function used in other place in this module.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 18 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: I remember that I moved it here is for debugging, and forgot to put it back.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 17 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs:37]: I will just drop `GetTogglePowerChanging` call here. I feel it senseless even for the mod that will have crazy idea on power mutiplier.

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 19 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: It is indeed a simple consideration here. But if we consider the multipliers change during the powedown time before toggling online, we'd better to do a power check to all just before we toggle on. However,

`GetTogglePowerChanging` itself is not very precise, because it will count all Power traits that enabled for actor based power, and all PowerMultiplier as traits that controlled by powerdown, which is compatible with default mods and SP/RV/CA, but if modder/game devs use PowerMultiplier as kinda upgrade or use Power/PowerMultiplier in more unexpected way, `GetTogglePowerChanging` will be dysfunctional.

In conclusion, the `toggledBuildings` is just a power change prediction, and at least for mod it support if the actual "powerdown" control cannot get to the correct answer in game, it can do it in next check.

File: OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs (hunk 15 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: And yeah we can replace 

```
			enabled = false;
			SetCondition(self, false);
```

as

```
TraitDisabled(self)
```

File: OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs (hunk 19 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs]: better idea on calculating power change of powerdown is welcomed here.

File: OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs (hunk 15 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: No, I was more wondering if this doesn't make more sense to be called from `INotifyRemovedFromWorld` or if that already calls `TraitDisabled` (which it apparently doesn't).

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: Ummm... 
Then I don't know that? I am not familiar with this interface.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleConditionOnOrder.cs]: Ummm... 
Then I don't know that? I am not familiar with this interface.

**Comment by pchote** [comment]
I've noted this PR in #19247 to track this for the future, but for now are closing this to help address the unsustainable state of the PR queue.
